site/src/pages/RoadmapPage/sidebar/MajorRequiredCourseList.tsx

```diff
@@ -74,105 +76,106 @@ const MajorRequiredCourseList: FC = () => {
-      updateSelectedMajorAndSpecialization(activePlanID, [
-        { majorId: selectedMajor.id, specializationId: specialization?.id },
-      ]);
-    },
-    [activePlanID, selectedMajor, savedSpecID, isLoggedIn],
-  );
+        if (specializationId) {
+          const specReqs = await getCoursesForSpecialization(specializationId);
+          allReqs = majorReqs.concat(specReqs);
+        }
 
-  // Spec or Major without specs selected, fetch requirements
-  useEffect(() => {
-    if (!selectedMajor?.id) return;
-    if (selectedMajor.specializations.length && !selectedSpec) return;
-    if (requirements.length) return;
-
-    setResultsLoading(true);
-    getCoursesForMajor(selectedMajor.id).then(async (majorReqs) => {
-      if (!selectedSpec) {
-        dispatch(setRequirements(majorReqs));
-        setResultsLoading(false);
-        return;
+        dispatch(setRequirements({ majorId, requirements: allReqs }));
+        if (!specsLoading[majorId] && !specOptions[majorId]) {
+          // console.log('fetching specs');
+          setSpecsLoading((prev) => ({ ...prev, [majorId]: true }));
+          const specs = await getMajorSpecializations(majorId);
+          specs.forEach((s) => (s.name = normalizeMajorName(s)));
+          specs.sort((a, b) => a.name.localeCompare(b.name));
+          const options = specs.map((s) => ({
+            value: s,
+            label: s.name,
+          }));
+          setSpecOptions((prev) => ({
+            ...prev,
+            [majorId]: options,
+          }));
+          setSpecsLoading((prev) => ({ ...prev, [majorId]: false }));
+        }
+      } finally {
+        setResultsLoading((prev) => ({ ...prev, [majorId]: false }));
       }
-      const specReqs = await getCoursesForSpecialization(selectedSpec.id);
-      dispatch(setRequirements(majorReqs.concat(specReqs)));
-      setResultsLoading(false);
-    });
-  }, [dispatch, selectedMajor, selectedSpec, requirements]);
+    },
+    [dispatch],
+  );
 
   // Switching roadmaps should restore major/spec
   useEffect(() => {
-    if (!majors.length || !activePlanID) return;
-    if (!activePlanID) return;
-    if (!isLoggedIn) return;
+    if (!majors.length || !activePlanID || !isLoggedIn) return;
 
+    let mounted = true;
     setMajorsLoading(true);
-    trpc.programs.getSavedMajorSpecPairs.query(activePlanID).then((pairs) => {
-      setMajorsLoading(false);
 
-      const foundMajor = majors.find((m) => m.id === pairs[0]?.majorId) ?? null;
-      dispatch(setMajor(foundMajor));
-      dispatch(setSpecialization(null));
-      setSavedSpecID(pairs[0]?.specializationId);
-      dispatch(setRequirements([]));
-    });
-  }, [dispatch, activePlanID, majors, isLoggedIn]);
+    const loadMajorsAndSpecs = async () => {
+      try {
+        const pairs = await trpc.programs.getSavedMajorSpecPairs.query(activePlanID);
+        if (!mounted) return; //mounted is false if the component is unmounted, prevents loop rendering issue i had
```

Comment:
This seems like it could be a structural issue. I think if each Major's spec/course list handles this on its own and calls the Redux store functions responsibly, you should be able to self-contain all of these instead of looping through pairs. Definitely easier said than done, but we could make some time to pair on this one.